Update: fix no-restricted-imports importNames reporting (fixes #12282)

fanich37 opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[X] Bug fix (it should fix #12282)
<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**
Now rule can produce multiple errors depends on `importNames` option. This change allows applying directive comments on each item imported from the module.

**Is there anything you'd like reviewers to focus on?**
Yes, the test cases since i'm too bad at it.

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -64,22 +64,26 @@ module.exports = {
                         },
                         additionalProperties: false
-                    }],
+                    },
```
These `[]` brackets were recently added in #12639 to avoid a possible misconfiguration. Did this change cause some other problems? Asking because you're removing them now, so I'm guessing there were some issues with the tests from this PR.

fanich37 replied:
Ooops... I think I lost it while rebasing...

fanich37 commented:
> I left some notes about `*` and `default`.
> 
> Perhaps we should also consider refactoring the rule.
> 
> The actual flow in this PR is:
> 
> 1. `checkNode()` collects `importNames` and calls `isRestrictedPath()` to find out is the import invalid.
> 2. `isRestrictedPath()` uses `importNames` and configuration to return `true` or `false`.
> 3. If it's `true`, `reportPath()` is called. Then, this function is supposed to report something, and do it with the precise locations.
> 
> The problem is - `reportPath()` doesn't get any useful info from the previous steps, apart from knowing that there's something invalid because it's being called? So, it basically repeats steps 1 and 2., meaning that these steps were almost useless. Maybe we should do everything in `checkNode()`, instead.

I refactored it a bit.

mdjermanovic commented:
Thanks for the changes!

I think there is no need to keep the existing structure of the code if it doesn't provide value for the new behavior.

In particular, it feels like `isRestrictedPath()` and `isRestrictedObject()` are not very useful now. I believe that if we add this simple `if` at the beginning of `reportPath()` in the actual version of this PR:

```js
function reportPath(importSource, importNames, node) {
    if (!restrictedPathMessages[importSource]) {
        return;
    }
    // ....
}
```
then we could rename `reportPath()`, call it directly and completely remove `isRestrictedPath()` and `isRestrictedObject()` functions from the code. Thoughts?

My idea was actually to simplify the code and do most of the logic in `checkNode()`:  in short, iterate over specifiers and report them right away (including `*`). Your a bit more complex approach with the map is also fine for me :+1:

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -248,26 +273,29 @@ module.exports = {
-            const importNames = node.specifiers ? node.specifiers.reduce((set, specifier) => {
+            const importNames = node.specifiers ? node.specifiers.reduce((map, specifier) => {
+                const specifierKey = specifier.imported ? "imported" : "local";
+                const specifierData = { loc: specifier.loc.start };
```
This should be just `specifier.loc`, without `.start`.

Also, it would be nice to have locations in the `invalid` tests (`column` and `endColumn` error properties, `line` if needed).

I think that it also makes sense to report the location of `*` in `reportPathForEverythingImported`.

fanich37 replied:
Done!

fanich37 commented:
> Thanks for the changes!
> 
> I think there is no need to keep the existing structure of the code if it doesn't provide value for the new behavior.
> 
> In particular, it feels like `isRestrictedPath()` and `isRestrictedObject()` are not very useful now. I believe that if we add this simple `if` at the beginning of `reportPath()` in the actual version of this PR:
> 
> ```js
> function reportPath(importSource, importNames, node) {
>     if (!restrictedPathMessages[importSource]) {
>         return;
>     }
>     // ....
> }
> ```
> 
> then we could rename `reportPath()`, call it directly and completely remove `isRestrictedPath()` and `isRestrictedObject()` functions from the code. Thoughts?
> 
> My idea was actually to simplify the code and do most of the logic in `checkNode()`: in short, iterate over specifiers and report them right away (including `*`). Your a bit more complex approach with the map is also fine for me 👍

Seems like a good idea! In this case I think that `isRestrictedForEverythingImported` and `reportPathForEverythingImported` are also useless. `*` could be reported in `reportPath`.
One thing I doubt about is reporting warnings in the specifiers loop. The clause to check whether report warning or not become quite big and difficult to reason about.

mdjermanovic commented:
> Seems like a good idea! In this case I think that `isRestrictedForEverythingImported` and `reportPathForEverythingImported` are also useless. `*` could be reported in `reportPath`.

I agree if it seems that the code would become easier to follow.

> One thing I doubt about is reporting warnings in the specifiers loop. The clause to check whether report warning or not become quite big and difficult to reason about.

Agreed, making the `importNames` map first (as it's implemented) seems like a better idea and a nice separation of logic :+1:

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -107,40 +116,66 @@ module.exports = {
-        function reportPath(node) {
-            const importSource = node.source.value.trim();
-            const customMessage = restrictedPathMessages[importSource] && restrictedPathMessages[importSource].message;
+        function reportPath(importSource, importNames, node) {
```
It would be nice to rename `reportPath` to `check...` something, as it's now more than just report.

fanich37 replied:
How about `checkRestrictedPathAndReport`?

mdjermanovic replied:
Sounds good! :+1:

fanich37 replied:
Done!

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -107,40 +116,66 @@ module.exports = {
+            const customMessage = restrictedPathMessages[importSource].message;
+            const restrictedImportNames = restrictedPathMessages[importSource].importNames;
+
+            if (importNames.has("*")) {
```
Should this be inside ` if (restrictedImportNames)`?

If the whole module is restricted (not just some `importNames` from the module, i.e. configuration for that module doesn't have `importNames`), it will be reported below with `"path"` or `"pathWithCustomMessage"` messageId.

As it is now in this PR, I think it would report two messages: one for the whole module (which is correct) and another for `*` (which shouldn't be reported when the whole module is restricted; also, the message for `*` expects `importNames` which doesn't exists in this case). It would be nice to add a test for this.

fanich37 replied:
Good catch! You're totally right 👍

fanich37 replied:
Done!

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -107,40 +116,66 @@ module.exports = {
+                    node,
+                    messageId: customMessage ? "everythingWithCustomMessage" : "everything",
+                    loc: data.loc,
+                    data: {
```
Might be nice to rename `data` variable (e.g. to `specifierData` as it's in `checkNode`) since these two unrelated `data`s together look a bit confusing?

fanich37 replied:
Completely agree. It's confusing. I even thought about renaming it but something went wrong.

fanich37 replied:
Done!

fanich37 commented:
> This looks very good! I just found a couple of minor details and I think there is one possible regression.

Thank you for review and suggestions. And sorry for the long journey with this PR :)

mdjermanovic commented:
Thanks for making all the changes and quickly addressing all concerns! It turned out that fixing this issue requires a quite big diff in the code.

As for the title of this PR, it might be safe to mark this as a semver-minor change (`Update:` instead of `Fix:`), because it can produce more warnings in some cases, caused by locations of `eslint-disable-*`  comments as we are now changing reported locations. Also, 'handle multiple errors' maybe doesn't quite describe this change.

Thoughts about something like this?

```
Update: fix no-restricted-imports importNames reporting (fixes #12282)
```

mdjermanovic commented:
About edge cases where the same name is imported multiple times in the same declaration, such as these:

```js
import { a, a as b } from "mod";

export { x as y, x as z } from "mod";

import foo, { default as bar } from "mod";
```

It might be correct to report all specifiers for a restricted name instead of just the last one, if it doesn't add too much complexity to the code. Maybe we could store arrays in `importNames` (e.g., find out `name` in the `if-else-if`s, then either `.set(name, [specifierData])` or `push` to existing). or do something else if this doesn't seem like a good idea.

It also seems that the configuration allows duplicates in `importNames`, but that should be a separate issue (it would be better to throw configuration error than to account for this in the code). It's okay for now to report the same specifier multiple times when the rule is misconfigured in such way.

fanich37 commented:
> Thanks for making all the changes and quickly addressing all concerns! It turned out that fixing this issue requires a quite big diff in the code.
> 
> As for the title of this PR, it might be safe to mark this as a semver-minor change (`Update:` instead of `Fix:`), because it can produce more warnings in some cases, caused by locations of `eslint-disable-*` comments as we are now changing reported locations. Also, 'handle multiple errors' maybe doesn't quite describe this change.
> 
> Thoughts about something like this?
> 
> ```
> Update: fix no-restricted-imports importNames reporting (fixes #12282)
> ```

Sounds good!

mdjermanovic commented:
Changed the title to `Update:...`

Regarding the cases where the same name is imported multiple times, I think it's fine to report just 1 specifier for now. Maybe we could change that later in a separate PR, if it looks useful.

fanich37 commented on tests/lib/rules/no-restricted-imports.js:
```diff
@@ -211,7 +269,13 @@ ruleTester.run("no-restricted-imports", rule, {
+            type: "ExportNamedDeclaration",
+            line: 1,
+            column: 9,
+            endColumn: 17
```
Could you please clarify the `endColumn` property. Should it be used from 'root' specifier or from imported/local?

mdjermanovic replied:
`column` and `endColumn` are just 1-based columns in the source code, unaware of the context. Also, `endColumn` has +1 added.

Or maybe I didn't understand the question. Did you mean should the error highlight `foo as b` or just `foo` in this particular example? Both choices make sense, I think it's ok to highlight `foo as b` as it is implemented.

fanich37 replied:
> Did you mean should the error highlight foo as b or just foo in this particular example?

That was my question exactly 👍

mdjermanovic replied:
This is how it looks in VSCode:

![image](https://user-images.githubusercontent.com/44349756/71936359-7deadf00-31a9-11ea-88c3-029fb1cacf8b.png)

I think it's okay. It would have made sense to report just `foo` as well, but then we should probably report just `*` in `* as foo`, and that would add some extra code to find the token, so I believe it isn't necessary.

mdjermanovic commented on tests/lib/rules/no-restricted-imports.js:
```diff
@@ -414,7 +520,135 @@ ruleTester.run("no-restricted-imports", rule, {
+            column: 8,
+            endColumn: 11
+        }, {
+            message: "* import is invalid because 'default' from 'mod' is restricted.",
```
This also covers a missing test for `"everything"` messageId. :+1:

fanich37 commented:
> About edge cases where the same name is imported multiple times in the same declaration, such as these:
> 
> ```js
> import { a, a as b } from "mod";
> 
> export { x as y, x as z } from "mod";
> 
> import foo, { default as bar } from "mod";
> ```
> 
> It might be correct to report all specifiers for a restricted name instead of just the last one, if it doesn't add too much complexity to the code. Maybe we could store arrays in `importNames` (e.g., find out `name` in the `if-else-if`s, then either `.set(name, [specifierData])` or `push` to existing). or do something else if this doesn't seem like a good idea.
> 
> It also seems that the configuration allows duplicates in `importNames`, but that should be a separate issue (it would be better to throw configuration error than to account for this in the code). It's okay for now to report the same specifier multiple times when the rule is misconfigured in such way.

The problem here is the clause `if (specifier.imported) ... else if (specifier.local)`. Since if specifier has `imported` property we never get into `local` branch of the clause...

mdjermanovic commented:
> The problem here is the clause `if (specifier.imported) ... else if (specifier.local)`. Since if specifier has `imported` property we never get into `local` branch of the clause...

I don't think it would be a problem. If there is `imported`, it really shouldn't take `local`.

In the `if-else-if` chain, `imported` is for specifiers that appear in `import` declarations, while the 
`local` is (probably unintuitive) for those in `export` declarations.

In the above examples:

```js
import { a, a as b } from "mod";

export { x as y, x as z } from "mod";
```

`import` has 2 specifiers, both have `a` as `imported`.

`export` has also 2 specifiers, both have `x` as `local`, and nothing as `imported` (`y` and `z` are `exported`).

[AST explorer link with the examples](https://astexplorer.net/#/gist/679506e5512d9abf9ab159ec2681071f/c19226b24577b263616cf1a06b2c1b5360678f18)

What I was thinking is, for example in the first case if `a` is a restricted name from `mod`, that it might be correct to report 2 errors. One error with the location of  the first `a` specifier, the other with the location of the `a as b` specifier. At the moment, `a as b` specifier will overwrite already set `a` location in the map. A solution might be to have `[{ loc }, { loc } ...]` in the map instead of just a single `{ loc }`.

fanich37 commented:
> This looks great, thanks for the fix and the refactoring!

Sorry for the delay. It's just new year holidays that is over and I've got a big task at work :) But I would like to add your proposal about duplicate import names. I think I could add it till Monday.

And also don't you think it would be better to put all logic from `checkPathAndReport` into `checkNode`, `checkPathAndReport` rename back into `reportPath`. I just think it's a bit weird we have two `check...` functions in a rule? 😕

mdjermanovic commented:
> Sorry for the delay. It's just new year holidays that is over and I've got a big task at work :) But I would like to add your proposal about duplicate import names. I think I could add it till Monday.

That would be great, thanks!

> And also don't you think it would be better to put all logic from `checkPathAndReport` into `checkNode`, `checkPathAndReport` rename back into `reportPath`. 

Something like my initial idea? I'm not sure about that anymore, the way it's implemented now looks very nice (i.e. deal with different kinds of imports/exports first to get the `importNames` map and then do the main logic on a simplified input).

Also, it looks like the actual version in this PR might make it easier to add some new things in the future, like to fix #12737 and support #12629. At least I think so, could be wrong. If you think that the other approach might be better for future maintenance and easier to follow, I'm fine with both.

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -107,40 +116,71 @@ module.exports = {
+
+                        if (importSource === "moda") {
+                            console.log("+++++", specifiers); // eslint-disable-line no-console
+                        }
```
This is probably left from debugging :)

fanich37 replied:
Ooops... sorry, forgot to add changes to the stage 🤦‍♂

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -248,26 +219,38 @@ module.exports = {
+                    } else {
+                        map.set(specifier.local.name, [specifierData]);
+                    }
                 }
```
This works fine! But wondering would it be easier to follow if we would just get the `name` in `if-else-if`s and then deal with the map after the chain? Then we wouldn't need to duplicate code for `imported` and `local` and think about why is `default` initialized as an array while `*` isn't. 

Code that checks `*` in `checkRestrictedPathAndReport` could take just `[0]`.

Thoughts?

fanich37 replied:
In this way it looks much more clearer and no duplication of the code. Awesome! Thank you!

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -248,26 +216,31 @@ module.exports = {
+                    map.set(name, map.get(name).concat(specifierData));
+                } else {
+                    map.set(name, [specifierData]);
+                }
```
Maybe we should wrap this in `if (name) {}`.

mdjermanovic replied:
An alternative for the `concat` line is `map.get(name).push(specifierData);` without `set`, though this also looks fine to me!

fanich37 replied:
> Maybe we should wrap this in `if (name) {}`.

Yes, you're right. You even mentioned above that could be the case where specifier might not have `imported` or `local` property 👍

fanich37 replied:
> An alternative for the `concat` line is `map.get(name).push(specifierData);` without `set`, though this also looks fine to me!

Yep, `push` is much faster 👍

mdjermanovic commented on lib/rules/no-restricted-imports.js:
```diff
@@ -107,40 +116,68 @@ module.exports = {
         /**
          * Report a restricted path.
+         * @param {string} importSource path of the import
+         * @param {Map<string,Object>} importNames Map of import names that are being imported
```
A minor detail, the map has arrays of objects instead of objects now.

fanich37 replied:
Good catch 👍

kaicataldo commented on lib/rules/no-restricted-imports.js:
```diff
@@ -95,6 +99,11 @@ module.exports = {
         const restrictedPatterns = (isPathAndPatternsObject ? options[0].patterns : []) || [];
 
+        // if no imports are restricted we don"t need to check
+        if (Object.keys(restrictedPaths).length === 0 && restrictedPatterns.length === 0) {
```
:+1:

fanich37 commented:
> Thanks, looks very good! Just two small details.

Thank you for your review and suggestions 👍

mdjermanovic commented:
@fanich37 thanks again for the really good work and I hope to see more PRs from you in the future! :+1:
